app/services/creator_analytics/churn.rb

```diff
@@ -0,0 +1,273 @@
+# frozen_string_literal: true
+
+class CreatorAnalytics::Churn
+  include ActiveModel::Validations
+
+  validates :end_date, comparison: { greater_than: :start_date }
+  validates :time_window, inclusion: { in: 1..31, message: "must be between 1 and 31 days" }
+
+  attr_reader :start_date, :end_date, :user, :products
+
+  def initialize(user:, start_date: nil, end_date: nil, params: {}, products: nil)
+    @user = user
+    @params = params
+    @start_date = parse_date(start_date, :start)
+    @end_date = parse_date(end_date, :end)
+    @products = products || parse_products
+  end
+
+  def time_window
+    (end_date - start_date).to_i + 1
+  end
+
+  def calculate
+    {
+      date: end_date,
+      customer_churn_rate: customer_churn_rate,
+      churned_subscribers: churned_count,
+      churned_mrr_cents: churned_mrr
+    }
+  end
+
+  def fetch_churn_data
+    return nil unless has_subscription_products?
+
+    if should_use_cache?
+      fetch_cached_data
+    else
+      fetch_realtime_data
+    end
+  end
+
+  def has_subscription_products?
+    user.products.alive.is_recurring_billing.exists?
+  end
+
+  def available_products
+    user.products_for_creator_analytics.select(&:is_recurring_billing?)
+  end
+
+  def calculate_by_date
+    earliest_date = start_date - 29.days
+    all_subscriptions = fetch_subscriptions(from: earliest_date, to: end_date)
+
+    (start_date..end_date).map do |date|
+      period_start = date - 29.days
+      calculate_for_period(period_start, date, all_subscriptions)
+    end
+  end
+
+  def customer_churn_rate
+    return 0.0 if total_subscriber_base.zero?
+    (churned_count.to_f / total_subscriber_base * 100).round(2)
+  end
```

Comment:
I think you could just reuse `calculate` here, can't you?
```suggestion
  def customer_churn_rate
    calculate[:customer_churn_rate]
  end
```